contracts/anchor/Anchor.sol

```diff
@@ -0,0 +1,207 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// ----------------------------------------------------------------------------
+//
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+import "./AnchorI.sol";
+
+import "../consensus/ConsensusModule.sol";
+import "../lib/CircularBufferUint.sol";
+import "../lib/MerklePatriciaProof.sol";
+import "../lib/RLP.sol";
+
+/**
+ * @title Anchor contract which implements StateRootInterface.
+ *
+ * @notice Anchor stores another chain's state roots. It stores the address of
+ *         the co-anchor, which will be the anchor on the other chain. State
+ *         roots are exchanged bidirectionally between the anchor and the
+ *         co-anchor by the consensus.
+ */
+contract Anchor is AnchorI, ConsensusModule, CircularBufferUint {
+
+    /* Usings */
+
+    using SafeMath for uint256;
+
+
+    /* Events */
+
+    event StateRootAvailable(uint256 _blockHeight, bytes32 _stateRoot);
+
+
+    /* Storage */
+
+    /** Maps block heights to their respective state root. */
+    mapping (uint256 => bytes32) private stateRoots;
+
+    /**
+     * The remote chain ID is the remote chain id where anchor contract is
+     * deployed.
+     */
+    uint256 private remoteChainId;
+
+    /** Address of the anchor on the auxiliary chain. Can be zero. */
+    address public coAnchor;
+
+
+    /*  Constructor */
+
+    /**
+     * @notice Contract constructor.
+     *
+     * @param _remoteChainId The chain id of the chain that is tracked by this
+     *                       anchor.
+     * @param _blockHeight Block height at which _stateRoot needs to store.
+     * @param _stateRoot State root hash of given _blockHeight.
+     * @param _maxStateRoots The max number of state roots to store in the
+     *                       circular buffer.
+     * @param _consensus A consensus address.
+     */
+    constructor(
+        uint256 _remoteChainId,
+        uint256 _blockHeight,
+        bytes32 _stateRoot,
+        uint256 _maxStateRoots,
+        address _consensus
+    )
+        ConsensusModule(_consensus)
+        CircularBufferUint(_maxStateRoots)
+        public
+    {
+        require(
+            _remoteChainId != 0,
+            "Remote chain Id must not be 0."
+        );
+
+        remoteChainId = _remoteChainId;
+
+        stateRoots[_blockHeight] = _stateRoot;
+        CircularBufferUint.store(_blockHeight);
+    }
+
+
+    /* External functions */
+
+    /**
+     *  @notice The Co-Anchor address is the address of the anchor that is
+     *          deployed on the other (origin/auxiliary) chain.
+     *
+     *  @param _coAnchor Address of the Co-Anchor on auxiliary.
+     */
+    function setCoAnchorAddress(address _coAnchor)
```

Comment:
we should probably get rid of this function; with the precompiles these contracts on the auxiliary chain would get a predetermined address

Reply:
Should we do it now or after having precompiles?

Reply:
we can get rid of it now, it s unused

Reply:
Done.